src/backend/catalog/test/storage_pending_deletes_redo_test.c

```diff
@@ -0,0 +1,1103 @@
+			}
+		case 16:
+			{
+				/* return not 0 size, but the returned pointer is NULL */
+				*size = 1;
+				return NULL;
+			}
+		default:
+			break;
+	}
+
+	/* return something valid */
+	int			node_count = 1;
+
+	*size = sizeof(size_t) + sizeof(PendingRelXactDelete) * (node_count);
+	char	   *buffer = palloc(*size);
+
+	PendingRelXactDeleteArray *pending_deletes =
+	(PendingRelXactDeleteArray *) buffer;
+
+	pending_deletes->count = node_count;
+
+	PendingRelXactDelete *pd = &(pending_deletes->array[0]);
+
+	pd->xid = TEST_XID;
+	pd->relnode.isTempRelation = false;
+	pd->relnode.node.spcNode = TEST_TABLESPACE_OID1;
+	pd->relnode.node.dbNode = TEST_DB_OID1;
+	pd->relnode.node.relNode = TEST_REL_OID1;
+
+	return pending_deletes;
+}
+
+XLogRecPtr
+__wrap_XLogInsert(RmgrId rmid, uint8 info, XLogRecData *rdata)
+{
+	assert_int_equal(test_number, 17);	/* currently we should get here only
+										 * in test_17 */
+	XLogInsert_call_count++;
+
+	assert_int_equal(rmid, RM_XLOG_ID);
+	assert_int_equal(info, XLOG_PENDING_DELETE);
+	assert_int_equal(rdata->buffer, InvalidBuffer);
+	assert_false(rdata->buffer_std);
+	assert_true(rdata->next == NULL);
+	assert_true(rdata->len == (sizeof(size_t) + sizeof(PendingRelXactDelete)));
+
+	PendingRelXactDeleteArray *pending_deletes =
+	(PendingRelXactDeleteArray *) rdata->data;
+
+	assert_int_equal(pending_deletes->count, 1);
+
+	PendingRelXactDelete *pd = &(pending_deletes->array[0]);
+
+	assert_int_equal(pd->xid, TEST_XID);
+	assert_false(pd->relnode.isTempRelation);
+	assert_int_equal(pd->relnode.node.spcNode, TEST_TABLESPACE_OID1);
+	assert_int_equal(pd->relnode.node.dbNode, TEST_DB_OID1);
+	assert_int_equal(pd->relnode.node.relNode, TEST_REL_OID1);
+
+	return TEST_XLOG_REC_PTR;
+}
+
+void
+__wrap_XLogFlush(XLogRecPtr record)
+{
+	assert_int_equal(test_number, 17);	/* currently we should get here only
+										 * in test_17 */
+	assert_int_equal(record, TEST_XLOG_REC_PTR);
+}
+
+/*
+ * Tests
+ */
+
+/*
+ * Scenario:
+ * add single pending delete node
+ * and then drop files.
+ */
+static void
+test_1(void **state)
+{
+	setup(1);
+	ShmemVariableCache->oldestXid = (TransactionId) 1;
+
+	PendingRelXactDelete pd = {0};
+
+	pd.xid = (TransactionId) 1;
+	pd.relnode.isTempRelation = false;
+	pd.relnode.node.spcNode = TEST_TABLESPACE_OID1;
+	pd.relnode.node.dbNode = TEST_DB_OID1;
+	pd.relnode.node.relNode = TEST_REL_OID1;
+
+	PdlRedoAdd(&pd);
+
+	PdlRedoDropFiles();
+
+	assert_int_equal(DropRelationFiles_call_count, 1);
+}
+
+/*
+ * Scenario:
+ * add single pending delete node
+ * and datfrozenxid is above the node's xid
+ * and then drop files.
+ */
+static void
+test_2(void **state)
+{
+	setup(2);
+	ShmemVariableCache->oldestXid = (TransactionId) 2;
+
+	PendingRelXactDelete pd = {0};
+
+	pd.xid = (TransactionId) 1;
+	pd.relnode.isTempRelation = false;
+	pd.relnode.node.spcNode = TEST_TABLESPACE_OID1;
+	pd.relnode.node.dbNode = TEST_DB_OID1;
+	pd.relnode.node.relNode = TEST_REL_OID1;
```

Comment:
Here and below I suggest to initialize pd in one go. PendingRelXactDelete pd = { .xid = (TransactionId) 1, ...}